074-k8s-storage/proposal.md

```diff
@@ -0,0 +1,167 @@
+
+## Requirements
+An ideal storage solution can do the following :
+
+- image fetching from the CRI k8s is using
+- transfer **cache objects** between steps (whatever represents a step, most likely a pod)
+- cache for resources and tasks
+- stream **cache objects** across worker runtimes (k8s worker sends artifact to garden worker)
+
+## Criteria
+- security
+- performance, aka initialization time (time spent running workloads on a single k8s worker, as well as across workers)
+- resource usage to run this storage solution
+
+# Proposal
+
+**TL;DR**: We recommend going with the image registry option because it satisfies all the requirements and gives us a bunch of options to improve performance when compared to the blobstore option. It also provides a very flexible solution that works across multiple runtime workers. [See Details](#image-registry-to-store-artifacts)
+
+Furthermore, the CSI is a useful interface for building the storage component against. [See Details](#csi)
+
+# Storage Options considered
+## Baggageclaim Daemonset 
+### Description
+A privileged baggageclaim pod would manage all the **cache object** for step pods. The pod can be provided sufficient privilege to create overlay mounts using `BiDirectional` value for `mountPropagation`. The `volumeMount` object allows specifying a volume `subPath`.
+
+This approach didn't work using GCE PDs or vSphere Volumes ([Issue](https://github.com/kubernetes/kubernetes/issues/95049)). It does work using `hostPath` option, however, that would require a large root volume and wouldn't be able to leverage IaaS based persistent disks. 
+
+The pod would run on all nodes that Concourse would execute steps on.
+ 
+### Pros
++ Leverage baggageclaim
+	+ volume streaming between nodes would work using the current Concourse architecture
+	+ resource and task caches would also work using the current Concourse architecture
+	+ would be able to stream **cache objects** across worker runtimes as it would be mediated via the web
++ Concourse would have complete control over volume lifecycle
++ would have negligible overhead for steps scheduled on the same node as no input/output stream would be required
+
+### Cons
+- Not being able to use IaaS based persisent disks doesn't offer a viable solution. K8s nodes would need to have large root volumes.
+- Wouldn't have support for hosting images by default. However, `baggageclaim` could be extended to add the APIs
+- `baggageclaim` itself doesn't have any authentication/authorization or transport security (https) mechanisms built into it
+
```

Comment:
```suggestion
+ - `baggageclaim` needs to be run in a privileged container. It would be useful to understand the perspective of operators for this constraint ? Is this tenable ?
```

Reply:
as far as i know, TKG clusters created via TMC expects pods to adhere to a podsecurity policy that's quite restrictive:

- https://docs.vmware.com/en/VMware-vSphere/7.0/vmware-vsphere-with-tanzu/GUID-CD033D1D-BAD2-41C4-A46F-647A560BAEAB.html
- https://docs.pivotal.io/tkgi/1-9/pod-security-policy.html

buuut, I'd imagine someone installing a CSI plugin (like anything network related) would expect to grant higher privileges anyway (being an "infra component")

Reply:
Yep, this does make the assumption that installing Concourse requires more than just pod* capabilities. 
However, installing a `CSI driver` would be an `Operator` action similar to installing concourse today.  

@taylorsilva and I were also discussing that this approach wouldn't preclude us from using the IaaS PVs directly for simpler installations where user doesn't have privileges to install a `CSI driver`. This would have performance implications in the overhead to create/destroy volumes and limit of volumes / node, but would be a very simple solution 🤔